### PR TITLE
[UWP] Fixes to Cell for ListView and CollectionView

### LINKED ISSUE #10473: [Bug] UWP - ListView does not recycle cells [HUUGE MEMORY ISSUE]
### Description
On UWP, the viewcells of a listview are **not** recycled. Rather, every time the datacontext changes, a new cell is created from the datatemplate!

### Steps to Reproduce

1. Create a page with a Xamarin.ListView and set the itemssource to 200 items
2. Create a custom ViewCell (so that you have the xaml and the code behind)
3. Set the `HeightRequest` of the viewcell to 80 or more
4. Set a breakpoint in the `ctor` of your ViewCell
5. Scroll all the way up and all the way down

### Expected Behavior
Constructor of the custom `ViewCell` should **never** be called during scrolling

### Actual Behavior
Constructor is called all the time...

### Basic Information

- Version with issue: 4.4.0 and all above
- Last known good version: no idea...
- IDE: VS 2019
- Platform Target Frameworks: 
   - UWP: irrelevant  
- Android Support Library Version: 
- Nuget Packages: irrelevant
- Affected Devices: all UWP devices

### Solution
We already found the issue for this.
The culprit is the `Xamarin.Forms.Platform.UAP.CellControl.SetCell(..)` method:

The error is marked with a ⚠
```csharp
void SetCell(object newContext)
{
	var cell = newContext as Cell;

	if (cell != null)
	{
		Cell = cell;
		return;
	}

	if (ReferenceEquals(Cell?.BindingContext, newContext))
		return;

	// If there is a ListView, load the Cell content from the ItemTemplate.
	// Otherwise, the given Cell is already a templated Cell from a TableView.
	ListView lv = _listView.Value;
	if (lv != null)
	{
		bool isGroupHeader = IsGroupHeader;
		DataTemplate template = isGroupHeader ? lv.GroupHeaderTemplate : lv.ItemTemplate;
		object bindingContext = newContext;

		if (template is DataTemplateSelector)
		{
			template = ((DataTemplateSelector)template).SelectTemplate(bindingContext, lv);
		}
                        // ⚠ ERROR: The code should now check if there already exists a Cell that
                        // could be recycled!
                        // Instead, however, it just creates a new cell using the template or the defaultcell (in case it is a groupheader)
		if (template != null)
		{
			cell = template.CreateContent() as Cell;
		}
		else
		{
			if (isGroupHeader)
				bindingContext = lv.GetDisplayTextFromGroup(bindingContext);

			cell = lv.CreateDefaultCell(bindingContext);
		}

		// A TableView cell should already have its parent,
		// but we need to set the parent for a ListView cell.
		cell.Parent = lv;

		// Set inherited BindingContext after setting the Parent so it won't be wiped out
		BindableObject.SetInheritedBindingContext(cell, bindingContext);

		// This provides the Group Header styling (e.g., larger font, etc.) when the
		// template is loaded later.
		cell.SetIsGroupHeader<ItemsView<Cell>, Cell>(isGroupHeader);
	}

	Cell = cell;
}
```

Now we patched this ourselves like the following:
(We marked our additions with rockeds: 🚀)
```csharp

void SetCell(object newContext)
{
	var cell = newContext as Cell;

	if (cell != null)
	{
		Cell = cell;
		return;
	}

	if (ReferenceEquals(Cell?.BindingContext, newContext))
		return;

	// If there is a ListView, load the Cell content from the ItemTemplate.
	// Otherwise, the given Cell is already a templated Cell from a TableView.
	ListView lv = _listView.Value;
	if (lv != null)
	{
		// 🚀 If there is an old cell, check if it was a group header
		// we need this later to know whether we can recycle this cell
		bool? wasGroupHeader = null;
		var oldCell = Cell;
		if (oldCell != null)
		{
			wasGroupHeader = oldCell.GetIsGroupHeader<ItemsView<Cell>, Cell>();
		}
		
		bool isGroupHeader = IsGroupHeader;
		DataTemplate template = isGroupHeader ? lv.GroupHeaderTemplate : lv.ItemTemplate;
		object bindingContext = newContext;

		bool sameTemplate = false;
		if (template is DataTemplateSelector dataTemplateSelector)
		{
			template = dataTemplateSelector.SelectTemplate(bindingContext, lv);

			 // 🚀 If there exists an old cell, get its data template and check
		        // whether the new- and old template matches. In that case, we can recycle it
		        if (oldCell?.BindingContext != null)
			{
				DataTemplate oldTemplate = dataTemplateSelector.SelectTemplate(oldCell?.BindingContext, lv);
				sameTemplate = oldTemplate == template;
			}
		}
		// 🚀 if there is no datatemplateselector, we now verify if the old cell
		// was a groupheader and whether the new one is as well.
		// Again, this is only to verify we can reuse this cell
		else if(wasGroupHeader.HasValue)
		{
			sameTemplate = wasGroupHeader == isGroupHeader;
		}

		// reuse cell
		var canReuseCell = Cell != null && sameTemplate;
		
		// 🚀 If we can reuse the cell, just reuse it...
		if (canReuseCell)
		{
			cell = Cell;
		}
		else if (template != null)
		{
			cell = template.CreateContent() as Cell;
		}
		else
		{
			if (isGroupHeader)
				bindingContext = lv.GetDisplayTextFromGroup(bindingContext);

			cell = lv.CreateDefaultCell(bindingContext);
		}

		// A TableView cell should already have its parent,
		// but we need to set the parent for a ListView cell.
		cell.Parent = lv;

		// Set inherited BindingContext after setting the Parent so it won't be wiped out
		BindableObject.SetInheritedBindingContext(cell, bindingContext);

		// This provides the Group Header styling (e.g., larger font, etc.) when the
		// template is loaded later.
		cell.SetIsGroupHeader<ItemsView<Cell>, Cell>(isGroupHeader);
	}

        // 🚀 Only set the cell if it DID change
	// Note: The cleanup (SendDisappearing(), etc.) is done by the Cell propertychanged callback so we do not need to do any cleanup ourselves.

	if(Cell != cell)
		Cell = cell;
	// 🚀 even if the cell did not change, we **must** call SendDisappearing() and SendAppearing()
	// because frameworks such as Reactive UI rely on this! (this.WhenActivated())
	else if(Cell != null)
	{
		Cell.SendDisappearing();
		Cell.SendAppearing();
	}

```

This works like a charm. 
I do not know where to target our PR though as there are so many versions affected.
Maybe you just copy paste the fix yourself in all platforms that apply?

@bruzkovsky  @BrayanKhosravian  - fyi ;-)

# Update:
It seems that the CellControl has another memory leak:
image

The `Cell.PropertyChanged` event is not unsubscribed when the cell is unloaded

```csharp

		public CellControl()
		{
			_listView = new Lazy<ListView>(GetListView);

			DataContextChanged += OnDataContextChanged;

			Unloaded += (sender, args) =>
			{
				Cell?.SendDisappearing();
			};

			_propertyChangedHandler = OnCellPropertyChanged;
		}
``` 

We'd fix it like this:

```csharp

		public CellControl()
		{
			_listView = new Lazy<ListView>(GetListView);

			DataContextChanged += OnDataContextChanged;

			Loaded += OnLoaded;
			Unloaded += OnUnloaded;

			_propertyChangedHandler = OnCellPropertyChanged;
		}

		void OnLoaded(object sender, RoutedEventArgs e)
		{
			if (Cell == null)
				return;

                        /// 🚀 subscribe topropertychanged
			// make sure we do not subscribe twice (because this could happen in SetSource(Cell oldCell, Cell newCell))
			Cell.PropertyChanged -= _propertyChangedHandler;
			Cell.PropertyChanged += _propertyChangedHandler;
		}

		void OnUnloaded(object sender, RoutedEventArgs e)
		{
			if (Cell == null)
				return;

			Cell.SendDisappearing();
                        /// 🚀 unsubscribe from propertychanged
			Cell.PropertyChanged -= _propertyChangedHandler;
		}

.....


		void SetSource(Cell oldCell, Cell newCell)
		{
			if (oldCell != null)
			{
				oldCell.PropertyChanged -= _propertyChangedHandler;
				oldCell.SendDisappearing();
			}

			if (newCell != null)
			{
				newCell.SendAppearing();

				UpdateContent(newCell);
				UpdateFlowDirection(newCell);
				SetupContextMenu();

				// 🚀 make sure we do not subscribe twice (OnLoaded!)
				newCell.PropertyChanged -= _propertyChangedHandler;
				newCell.PropertyChanged += _propertyChangedHandler;
			}
		}



```

### PR DESCRIPTION
**Pull request by CliffAgius**
### Description of Change ###

Fix for #10473 that is a massive blocker for UWP development using List and Collection views in XF.  This fix was designed by @gentledepp and I am putting it forward as a PR to the main XF repo as they had only made the changes to thier company fork.  I have tested these changes and they seem to fix the issue on the project we are building.

### Issues Resolved ### 
- fixes #10473 

### API Changes ###
None

### Platforms Affected ### 
- UWP

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -27,34 +27,6 @@ resources:
       name: xamarin/yaml-templates
       endpoint: xamarin
 
-trigger:
```
Comment (rmarinho) [inline azure-pipelines.yml:30]: please remove these changes

Comment (CliffAgius) [reply azure-pipelines.yml:30]: Odd I havn't touched that file... Ok will check and update standby.

**Comment by CliffAgius** [comment]
here you go @rmarinho I reverted the changes I have no idea how or why that file was changed but it's back to how it was and only the one file is now changed which is the one that should I hope fix the UWP Memory leak.

**Review by hartez — APPROVED** [review]

**Pull request merged**